File: app/src/main/java/com/liuzhenli/reader/ui/fragment/DiscoverFragment.java
package com.liuzhenli.reader.ui.fragment;

import android.content.Context;
import android.graphics.Typeface;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AccelerateInterpolator;
import android.view.animation.DecelerateInterpolator;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentStatePagerAdapter;

import com.hwangjr.rxbus.RxBus;
import com.liuzhenli.common.constant.RxBusTag;
import com.liuzhenli.common.base.BaseFragment;
import com.liuzhenli.common.AppComponent;
import com.liuzhenli.reader.DaggerReadBookComponent;
import com.liuzhenli.reader.ui.contract.DiscoverContract;
import com.liuzhenli.reader.ui.presenter.DiscoverPresenter;
import com.liuzhenli.reader.view.ScaleTransitionPagerTitleView;
import com.micoredu.reader.analyzerule.AnalyzeRule;
import com.micoredu.reader.bean.BookCategoryBean;
import com.micoredu.reader.bean.BookSourceBean;
import com.microedu.reader.R;
import com.microedu.reader.databinding.FragmentDiscoverBinding;


import net.lucode.hackware.magicindicator.ViewPagerHelper;
import net.lucode.hackware.magicindicator.buildins.UIUtil;
import net.lucode.hackware.magicindicator.buildins.commonnavigator.CommonNavigator;
import net.lucode.hackware.magicindicator.buildins.commonnavigator.abs.CommonNavigatorAdapter;
import net.lucode.hackware.magicindicator.buildins.commonnavigator.abs.IPagerIndicator;
import net.lucode.hackware.magicindicator.buildins.commonnavigator.abs.IPagerTitleView;
import net.lucode.hackware.magicindicator.buildins.commonnavigator.indicators.LinePagerIndicator;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import javax.script.ScriptException;
import javax.script.SimpleBindings;

import static androidx.fragment.app.FragmentPagerAdapter.BEHAVIOR_SET_USER_VISIBLE_HINT;
import static com.liuzhenli.common.constant.AppConstant.SCRIPT_ENGINE;

/**
 * describe:推荐书源列表
 *
 * @author Liuzhenli on 2019-11-09 22:28
 */
public class DiscoverFragment extends BaseFragment<DiscoverPresenter> implements DiscoverContract.View {

    private CommonNavigatorAdapter mCommonNavigationAdapter;
    private FragmentStatePagerAdapter fragmentPagerAdapter;
    private ArrayList<BookCategoryBean> mBookCategory = new ArrayList<>();
    protected List<BaseFragment> mFragmentList = new ArrayList<>();
    private FragmentDiscoverBinding inflate;

    public static DiscoverFragment getInstance() {
        DiscoverFragment instance = new DiscoverFragment();
        Bundle bundle = new Bundle();
        instance.setArguments(bundle);
        return instance;
    }

    @Override
    public View bindContentView(LayoutInflater inflater, ViewGroup container, boolean attachParent) {
        inflate = FragmentDiscoverBinding.inflate(inflater, container, attachParent);
        return inflate.getRoot();
    }

    @Override
    protected void setupActivityComponent(AppComponent appComponent) {
        DaggerReadBookComponent.builder().build().inject(this);
    }

    @Override
    public void attachView() {
        mPresenter.attachView(this);
    }


    @Override
    public void initData() {

    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mPresenter.detachView();
    }

    @Override
    public void configViews() {
        inflate.mViewPager.setOffscreenPageLimit(10);
        FragmentManager fm = getChildFragmentManager();
        fragmentPagerAdapter = new FragmentStatePagerAdapter(fm, BEHAVIOR_SET_USER_VISIBLE_HINT) {
            @Override
            public Fragment getItem(int position) {
                return mFragmentList.get(position);
            }

            @Override
            public int getCount() {
                return mFragmentList.size();
            }

            @Nullable
            @Override
            public CharSequence getPageTitle(int position) {
                return mBookCategory.get(position).name;
            }

            @Override
            public int getItemPosition(@NotNull Object object) {
                if (mFragmentList.contains((Fragment) object)) {
                    // 如果当前 item 未被 remove，则返回 item 的真实 position
                    return mFragmentList.indexOf((Fragment) object);
                } else {
                    // 否则返回状态值 POSITION_NONE
                    return POSITION_NONE;
                }
            }
        };
        inflate.mViewPager.setAdapter(fragmentPagerAdapter);

        CommonNavigator commonNavigator7 = new CommonNavigator(mContext);
        //这个控制左右滑动的时候,选中文字的位置,0.5表示在中间
        commonNavigator7.setScrollPivotX(0.5f);
        mCommonNavigationAdapter = new CommonNavigatorAdapter() {
            @Override
            public int getCount() {
                return mFragmentList.size();
            }

            @Override
            public IPagerTitleView getTitleView(Context context, int index) {
                ScaleTransitionPagerTitleView simplePagerTitleView = new ScaleTransitionPagerTitleView(context);
                simplePagerTitleView.setTypeface(Typeface.defaultFromStyle(Typeface.BOLD));
                simplePagerTitleView.setText(mBookCategory.get(index).name);
                simplePagerTitleView.setTextSize(16);
                simplePagerTitleView.setNormalColor(getResources().getColor(R.color.text_color_99));
                simplePagerTitleView.setSelectedColor(getResources().getColor(R.color.text_color_66));
                simplePagerTitleView.setOnClickListener(v -> inflate.mViewPager.setCurrentItem(index));
                return simplePagerTitleView;
            }

            @Override
            public IPagerIndicator getIndicator(Context context) {
                LinePagerIndicator indicator = new LinePagerIndicator(context);
                indicator.setMode(LinePagerIndicator.MODE_EXACTLY);
                indicator.setLineHeight(UIUtil.dip2px(context, 4));
                indicator.setLineWidth(UIUtil.dip2px(context, 10));
                indicator.setRoundRadius(UIUtil.dip2px(context, 3));
                indicator.setStartInterpolator(new AccelerateInterpolator());
                indicator.setEndInterpolator(new DecelerateInterpolator(2.0f));
                indicator.setColors(getResources().getColor(R.color.color_widget));
                return indicator;
            }
        }

        ;
        commonNavigator7.setAdapter(mCommonNavigationAdapter);
        inflate.mIndicator.setNavigator(commonNavigator7);
        ViewPagerHelper.bind(inflate.mIndicator, inflate.mViewPager);
        inflate.mBookSourceView.setOnItemClick(this::onBookSourceChange);
        inflate.mBookSourceView.setVisibility(View.GONE);
        TextView tvText = inflate.mViewEmpty.viewEmpty.findViewById(R.id.tv_empty_text);
        if (inflate.mViewEmpty != null) {
            tvText.setText("暂无书源,搜索微信公众号:异书房,\n回复\"书源\"获取书源~");
        }
    }

    public void onRefresh() {
        mPresenter.getSource();
    }

    @Override
    public void showSource(List<BookSourceBean> bookSourceData) {
        if (bookSourceData != null && bookSourceData.size() > 0) {
            Random random = new Random();
            int index = random.nextInt(bookSourceData.size());
            onBookSourceChange(bookSourceData.get(index));
            inflate.mViewEmpty.viewEmpty.setVisibility(View.GONE);
        } else {
            inflate.mViewEmpty.viewEmpty.setVisibility(View.VISIBLE);
        }
        inflate.mBookSourceView.setData(bookSourceData);
    }


    @Override
    public void showError(Exception e) {
    }

    @Override
    public void complete() {
    }

    @Override
    protected void onFragmentVisibleChange(boolean isVisible) {
        super.onFragmentVisibleChange(isVisible);
        if (isVisible) {
            onRefresh();
        }
    }

    public void showBookSourceView() {
        if (inflate == null) {
            return;
        }
        if (inflate.mBookSourceView.getVisibility() == View.GONE) {
            inflate.mBookSourceView.show();
        } else {
            inflate.mBookSourceView.close();
        }
    }


    public synchronized void onBookSourceChange(BookSourceBean bookSourceBean) {
        mFragmentList.clear();
        mBookCategory.clear();
        String ruleFindUrl = bookSourceBean.getRuleFindUrl();
        if (TextUtils.isEmpty(ruleFindUrl)) {
            return;
        }
        if (ruleFindUrl.startsWith("<js>")) {
            try {
                String jsStr = ruleFindUrl.substring(4, ruleFindUrl.lastIndexOf("<"));
                SimpleBindings bindings = new SimpleBindings();
                bindings.put("java", new AnalyzeRule(null, bookSourceBean));
                bindings.put("baseUrl", ruleFindUrl);
                Object object = SCRIPT_ENGINE.eval(jsStr, bindings);
                ruleFindUrl = object.toString();
            } catch (ScriptException e) {
                e.printStackTrace();
            }
        }
        //一个类目
        String[] findItem = ruleFindUrl.split("(&&|\n)+");
        for (String s : findItem) {
            String[] categoryInfo = s.split("::");
            if (categoryInfo.length < 2) {
                continue;
            }
            mBookCategory.add(new BookCategoryBean(categoryInfo[0], categoryInfo[1]));
            String url = categoryInfo[1];
            String tag = bookSourceBean.getBookSourceUrl();
            mFragmentList.add(BookCategoryFragment.getInstance(url, tag, bookSourceBean.getBookSourceName()));
        }
        fragmentPagerAdapter.notifyDataSetChanged();
        mCommonNavigationAdapter.notifyDataSetChanged();
        inflate.mViewPager.setCurrentItem(0);
        RxBus.get().post(RxBusTag.CHANGE_DISCOVER_BOOK_SOURCE, bookSourceBean);
    }

    @Override
    public void onResume() {
        super.onResume();
        //如果没有书源显示,再刷新一次
        if (inflate.mViewEmpty.viewEmpty.getVisibility() == View.VISIBLE) {
            onRefresh();
        }
    }


    public boolean isShowBookSourceMenu() {
        return inflate.mBookSourceView.isShowing();
    }

    public void dismissBookSourceMenu() {
        inflate.mBookSourceView.close();
    }

}

File: app/src/main/java/com/liuzhenli/reader/ui/discover/DiscoverFragment.kt
package com.liuzhenli.reader.ui.discover

import android.content.Context
import android.graphics.Typeface
import android.os.Bundle
import android.text.TextUtils
import android.view.View
import android.view.animation.AccelerateInterpolator
import android.view.animation.DecelerateInterpolator
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.viewpager2.adapter.FragmentStateAdapter
import androidx.viewpager2.widget.ViewPager2
import by.kirich1409.viewbindingdelegate.viewBinding
import com.airbnb.mvrx.MavericksView
import com.airbnb.mvrx.fragmentViewModel
import com.airbnb.mvrx.withState
import com.liuzhenli.common.base.BaseListener
import com.liuzhenli.reader.ui.discover.category.BookCategoryFragment
import com.liuzhenli.reader.view.ScaleTransitionPagerTitleView
import com.micoredu.reader.BaseFragment
import com.micoredu.reader.R
import com.micoredu.reader.bean.BookSource
import com.micoredu.reader.bean.rule.ExploreKind
import com.micoredu.reader.constant.EventBus
import com.micoredu.reader.databinding.FragmentDiscoverBinding
import com.micoredu.reader.utils.postEvent
import net.lucode.hackware.magicindicator.buildins.UIUtil
import net.lucode.hackware.magicindicator.buildins.commonnavigator.CommonNavigator
import net.lucode.hackware.magicindicator.buildins.commonnavigator.abs.CommonNavigatorAdapter
import net.lucode.hackware.magicindicator.buildins.commonnavigator.abs.IPagerIndicator
import net.lucode.hackware.magicindicator.buildins.commonnavigator.abs.IPagerTitleView
import net.lucode.hackware.magicindicator.buildins.commonnavigator.indicators.LinePagerIndicator


class DiscoverFragment : BaseFragment(R.layout.fragment_discover),
    MavericksView {
    val binding by viewBinding(FragmentDiscoverBinding::bind)
    private val mViewModel: DiscoverViewModel by fragmentViewModel()

    private var fragmentsList = mutableListOf<BookCategoryFragment>()
    private var pagerAdapter: FragmentStateAdapter? = null
    private var navigatorAdapter: CommonNavigatorAdapter? = null
    private var mBookCategory = mutableListOf<ExploreKind>()

    override fun invalidate() = withState(mViewModel) {

    }


    override fun init(savedInstanceState: Bundle?) {
        mViewModel.onAsync(
            DiscoverState::queryBookSource,
            deliveryMode = uniqueOnly("DiscoverFragment_queryBookSource"),
            onSuccess = { list ->
                if (list.isNotEmpty()) {
                    val bookSources = list.filterNot { source ->
                        TextUtils.isEmpty(source.exploreUrl)
                    }
                    binding.mBookSourceView.setData(bookSources)
                    val bookSource = bookSources.shuffled().take(1)[0]
                    mViewModel.dealCategory(bookSource)
                    postEvent(EventBus.CHANGE_DISCOVER_TITLE, bookSource.bookSourceName)
                }

            },
            onFail = {})

        mViewModel.onAsync(
            DiscoverState::dealCategory,
            deliveryMode = uniqueOnly("DiscoverFragment_dealCategory"),
            onSuccess = {
                onBookSourceChange(it)
            },
            onFail = {})
        binding.mViewPager.offscreenPageLimit = 10
        pagerAdapter = object : FragmentStateAdapter(requireActivity()) {

            override fun getItemCount(): Int {
                return fragmentsList.size
            }

            override fun createFragment(position: Int): Fragment {
                return fragmentsList[position]
            }

            override fun getItemId(position: Int): Long {
                return mBookCategory[position].hashCode().toLong()
            }

            override fun containsItem(itemId: Long): Boolean =
                mBookCategory.any { it.hashCode().toLong() == itemId }
        }

        binding.mViewPager.adapter = pagerAdapter
        val navigator = CommonNavigator(context)
        navigatorAdapter = object : CommonNavigatorAdapter() {
            override fun getCount(): Int {
                return mBookCategory.size
            }

            override fun getTitleView(context: Context?, index: Int): IPagerTitleView {
                val simplePagerTitleView = ScaleTransitionPagerTitleView(context)
                simplePagerTitleView.typeface = Typeface.defaultFromStyle(Typeface.BOLD)
                simplePagerTitleView.text = mBookCategory[index].title.replace("\\s".toRegex(), "")
                simplePagerTitleView.textSize = 16f
                simplePagerTitleView.normalColor = resources.getColor(R.color.text_color_99)
                simplePagerTitleView.selectedColor = resources.getColor(R.color.text_color_66)
                simplePagerTitleView.setOnClickListener { v: View? ->
                    binding.mViewPager.setCurrentItem(index, false)
                }
                return simplePagerTitleView
            }

            override fun getIndicator(context: Context?): IPagerIndicator {
                val indicator = LinePagerIndicator(context)
                indicator.mode = LinePagerIndicator.MODE_EXACTLY
                indicator.lineHeight = UIUtil.dip2px(context, 4.0).toFloat()
                indicator.lineWidth = UIUtil.dip2px(context, 10.0).toFloat()
                indicator.roundRadius = UIUtil.dip2px(context, 3.0).toFloat()
                indicator.startInterpolator = AccelerateInterpolator()
                indicator.endInterpolator = DecelerateInterpolator(2.0f)
                indicator.setColors(resources.getColor(R.color.color_widget))
                return indicator
            }
        }
        navigator.adapter = navigatorAdapter
//        navigator.isReselectWhenLayout = false
        binding.mIndicator.navigator = navigator

        binding.mViewPager.registerOnPageChangeCallback(object : ViewPager2.OnPageChangeCallback() {
            override fun onPageScrolled(
                position: Int,
                positionOffset: Float,
                positionOffsetPixels: Int
            ) {
                super.onPageScrolled(position, positionOffset, positionOffsetPixels)
                binding.mIndicator.onPageScrolled(position, positionOffset, positionOffsetPixels)
            }

            override fun onPageSelected(position: Int) {
                super.onPageSelected(position)
                binding.mIndicator.onPageSelected(position)
                binding.mIndicator.onPageScrolled(position, 0f, 0)
            }

            override fun onPageScrollStateChanged(state: Int) {
                super.onPageScrollStateChanged(state)
                binding.mIndicator.onPageScrollStateChanged(state)
            }
        })

        binding.mBookSourceView.setOnItemClick(object : BaseListener<BookSource> {
            override fun onResponse(bookSource: BookSource) {
                postEvent(EventBus.CHANGE_DISCOVER_TITLE, bookSource?.bookSourceName)
                mViewModel.dealCategory(bookSource!!)
            }
        })
        binding.mBookSourceView.visibility = View.GONE
        val tvText: TextView = binding.mViewEmpty.viewEmpty.findViewById(R.id.tv_empty_text)
        tvText.text = "暂无书源,搜索微信公众号:异书房,\n回复\"书源\"获取书源~"
        mViewModel.queryBookSource()
    }

    @Synchronized
    fun onBookSourceChange(list: List<ExploreKind>) {
        val categories = list.filterNot { TextUtils.isEmpty(it.url) }

        binding.mViewEmpty.viewEmpty.visibility = View.GONE

        fragmentsList.clear()
        mBookCategory.clear()

        mBookCategory.addAll(categories)
        for (s in categories) {
            val fragment = BookCategoryFragment()
            val bundle = Bundle()
            bundle.putSerializable("category", s)
            bundle.putSerializable("bookSourceUrl", mViewModel.bookSource?.bookSourceUrl)
            fragment.arguments = bundle
            fragmentsList.add(fragment)
        }
        pagerAdapter?.notifyDataSetChanged()
        navigatorAdapter?.notifyDataSetChanged()
        binding.mViewPager.currentItem = 0
    }


    fun configTitle() {
        if (binding.mBookSourceView.isClosed) {
            binding.mBookSourceView.show()
        } else {
            binding.mBookSourceView.close()
        }
    }
}